src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudInputFormat.java

```diff
@@ -0,0 +1,165 @@
+   * @return Collection of MarketingCloudObjectInfo containing table and schema.
+   */
+  public static List<MarketingCloudObjectInfo> setInput(Configuration jobConfig, SourceQueryMode mode,
+                                                        MarketingCloudSourceConfig conf) {
+    MarketingCloudJobConfiguration jobConf = new MarketingCloudJobConfiguration(jobConfig);
+    jobConf.setPluginConfiguration(conf);
+
+    //Depending on the selected objects in the conf, get the schema for each object as MarketingCloudObjectInfo
+    List<MarketingCloudObjectInfo> tableInfos = fetchTableInfo(mode, conf);
+
+    jobConf.setTableInfos(tableInfos);
+
+    LOG.debug("setInput::tableInfos = {}", tableInfos.size());
+    return tableInfos;
+  }
+
+  /**
+   * Depending on conf value fetch the list of fields for each object and create schema object.
+   *
+   * @param mode the query mode
+   * @param conf the plugin conf
+   * @return Collection of MarketingCloudObjectInfo containing table and schema.
+   */
+  private static List<MarketingCloudObjectInfo> fetchTableInfo(SourceQueryMode mode, MarketingCloudSourceConfig conf) {
+    try {
+      MarketingCloudClient client = MarketingCloudClient.create(conf.getClientId(), conf.getClientSecret(),
+                                                                conf.getAuthEndpoint(), conf.getSoapEndpoint());
+
+      //When mode = SingleObject, fetch fields for the object selected in plugin config
+      if (mode == SourceQueryMode.SINGLE_OBJECT) {
+        MarketingCloudObjectInfo tableInfo = getTableMetaData(conf.getObject(), conf.getDataExtensionKey(), client);
+        return (tableInfo == null) ? Collections.emptyList() : Collections.singletonList(tableInfo);
+      }
+
+      //When mode = MultiObject, get the list of objects provided in plugin config and the fetch fields for each of
+      //then objects.
+      List<MarketingCloudObjectInfo> tableInfos = new ArrayList<>();
+      List<SourceObject> objectList = conf.getObjectList();
+
+      for (SourceObject object : objectList) {
+        MarketingCloudObjectInfo tableInfo = null;
+
+        if (object == SourceObject.DATA_EXTENSION) {
+          //if the object = Data Extension then get the list of data extension keys and then fetch fields for each of
+          //the data extension objects.
+          List<String> dataExtensionKeys = Util.splitToList(conf.getDataExtensionKeys(), ',');
+          for (String dataExtensionKey : dataExtensionKeys) {
+            tableInfo = getTableMetaData(object, dataExtensionKey, client);
+            if (tableInfo == null) {
+              continue;
+            }
+            tableInfos.add(tableInfo);
+          }
+        } else {
+          tableInfo = getTableMetaData(object, "", client);
+          if (tableInfo != null) {
+            tableInfos.add(tableInfo);
+          }
+        }
+      }
+
+      return tableInfos;
+    } catch (Exception e) {
+      LOG.error("Error in fetchTableInfo()", e);
```

Comment:
This error message is not good. How should user react to this message. Could we have descriptive error message with additional information about how they can fix the problem.

Reply:
Added descriptive message.